File: P2/quicksort/quicksort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <sys/time.h>

void swap(int* a, int* b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

int partition(int arr[], int low, int high) {
    int pivot = arr[high];
    int i = (low - 1);
    
    for (int j = low; j <= high - 1; j++) {
        if (arr[j] < pivot) {
            i++;
            swap(&arr[i], &arr[j]);
        }
    }
    swap(&arr[i + 1], &arr[high]);
    return (i + 1);
}

void quickSort(int arr[], int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

int compare(const void* a, const void* b) {
    return (*(int*)a - *(int*)b);
}

double get_time() {
    struct timeval tv;
    gettimeofday(&tv, NULL);
    return tv.tv_sec + tv.tv_usec * 1e-6;
}

void test_sorting(int size) {
    int *arr1 = malloc(size * sizeof(int));
    int *arr2 = malloc(size * sizeof(int));
    
    for(int i = 0; i < size; i++) {
        arr1[i] = rand() % 10000;
        arr2[i] = arr1[i];
    }
    
    double start, end;
    
    // Quicksort Manual
    start = get_time();
    quickSort(arr1, 0, size - 1);
    end = get_time();
    printf("Quicksort Manual  (n=%d): %f segundos\n", size, end - start);
    
    // stdlib qsort
    start = get_time();
    qsort(arr2, size, sizeof(int), compare);
    end = get_time();
    printf("Stdlib Qsort (n=%d): %f segundos\n\n", size, end - start);
    
    free(arr1);
    free(arr2);
}

int main() {
    srand(time(NULL));
    
    int sizes[] = {10, 100, 1000, 10000, 100000, 1000000};
    int num_sizes = sizeof(sizes) / sizeof(sizes[0]);
    
    printf("Comparando algoritmos de ordenación...\n\n");
    
    for(int i = 0; i < num_sizes; i++) {
        test_sorting(sizes[i]);
    }
    
    return 0;
}

File: P2/quicksort/quicksort.java
import java.util.Arrays;
import java.util.Random;

public class quicksort {
    
    public static void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            int pi = partition(arr, low, high);
            quickSort(arr, low, pi - 1);
            quickSort(arr, pi + 1, high);
        }
    }
    
    private static int partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int i = low - 1;
        
        for (int j = low; j < high; j++) {
            if (arr[j] <= pivot) {
                i++;
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }
        
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;
        
        return i + 1;
    }
    
    public static void main(String[] args) {
        int[] sizes = {10, 100, 1000, 10000, 100000, 1000000};
        Random rand = new Random();
        
        System.out.println("Len\tQuickSort Manual\nArrays.sort");
        System.out.println("----------------------------------------");
        
        for (int size : sizes) {
            int[] arr1 = new int[size];
            int[] arr2 = new int[size];
            
            for (int i = 0; i < size; i++) {
                int num = rand.nextInt(1000000);
                arr1[i] = num;
                arr2[i] = num;
            }
            
            // quicksort manual
            long startTime = System.nanoTime();
            quickSort(arr1, 0, arr1.length - 1);
            long manualTime = System.nanoTime() - startTime;
            
            // Arrays.sort
            startTime = System.nanoTime();
            Arrays.sort(arr2);
            long javaTime = System.nanoTime() - startTime;
            
            System.out.printf("%d\t%.3f ms\t%.3f ms%n", 
                size, 
                manualTime / 1_000_000.0, 
                javaTime / 1_000_000.0);
        }
    }
}